webview: Add support for 'time' elements.

agrawal-d opened the pull request:
Add support for rendering 'time' mentions in message text. The
mentioned time is formatted to be human-readable, and tapping on it
shows an alert dialog with the original text.

**Screenshots**

![image](https://user-images.githubusercontent.com/7714968/86213101-cc949700-bb96-11ea-9e0d-e56554740938.png)

![image](https://user-images.githubusercontent.com/7714968/86214403-c6071f00-bb98-11ea-8ad6-05772f44bbb3.png)

![image](https://user-images.githubusercontent.com/7714968/86213067-be467b00-bb96-11ea-8529-e23a4bc5c880.png)



Fixes: #4162

agrawal-d commented:
The first commit 5f6b7a4 is borrowed from my PR https://github.com/zulip/zulip-mobile/pull/4169.

chrisbobbe commented on src/webview/webViewEventHandlers.js:
```diff
@@ -125,7 +130,8 @@ export type MessageListEvent =
   | MessageListEventError
-  | MessageListEventMention;
+  | MessageListEventMention
+  | MessageListTime;
```
Let's have this start with `MessageListEvent`, like the others. Then maybe have something a bit more descriptive; `MessageListEventTimeDetails` (looking at `MessageListEventReactionDetails`, above) or `MessageListEventPressTime`, maybe.

The names for the events going into and out of the WebView are very confusing; one part of #3913 is to fix that (see https://github.com/zulip/zulip-mobile/pull/3913/commits/6c9ff80864780c2e51164a642202d9de408d5ca5). But that PR has gone rather stale, and I'm hoping to get back to it before too long—likely after your PR lands, though.

chrisbobbe commented on src/webview/js/rewriteHtml.js:
```diff
@@ -58,11 +58,37 @@ export const rewriteImageUrls = (auth: Auth, element: Element | Document = docum
+    const timeStamp = elem.getAttribute('datetime');
+    const text = elem.innerText;
+
+    // 'text' and 'timeStamp' are guaranteed to be present.
```
This comment, and the code below it, look a little funny to me, like, "This will definitely be here. If it's not...".

I think the check is probably being done to satisfy Flow that these things will, in fact, be present. A brief explanation of that fact would help this make more sense, so it doesn't seem self-contradictory. 🙂

chrisbobbe commented on src/webview/js/rewriteHtml.js:
```diff
@@ -58,11 +58,37 @@ export const rewriteImageUrls = (auth: Auth, element: Element | Document = docum
+
+    const d = new Date(timeStamp);
+    elem.setAttribute('original-text', text);
+    elem.innerText = `🕒  ${d.toLocaleString(undefined, {
```
If you pass `undefined`, I wonder if it's smart enough to detect the correct locale, e.g., for putting the month before the day or vice versa: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat#Parameters

I think maybe it is; this is running in a WebView, which the device has some control over—hopefully enough to tell it the appropriate device-level setting.

agrawal-d commented:
Thanks for the review! I've pushed some changes. 
I've removed the `If no root element is specified, transform every <img> in the entire document.` comment from `rewriteImageUrls` because it's private, and is only called by `rewriteHtml`, which always passes an element or document to it - it's redundant and also incorrect.

Instead, as you suggested, I've added a similar comment to `rewriteHtml` instead.

chrisbobbe commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+ * Modify relevant parts of the rendered DOM tree under the given root,
+ * inclusive, to ensure functionality.
+ *
+ * If no root element is specified, the eintire document is considered.
```
nit: 'entire"

chrisbobbe commented on src/webview/js/__tests__/rewriteHtml.js:
```diff
@@ -1,7 +1,7 @@
 /** @jest-environment jsdom-global */
 // @flow strict-local
 
-import rewriteImageUrls from '../rewriteImageUrls';
```
fa7f7539a webview [nfc]: Refactor rewriteImageUrls to rewriteHtml.

Commit message nits: "Rename", "create".

agrawal-d replied:
I thought this form ("Renames") of commit message body is the one recommended by Zulip [Docs](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)?

>The body is written in prose, with full paragraphs; each paragraph should be separated from the next by a single blank line.

chrisbobbe replied:
Ah, I think the mobile project diverges from the server project in this respect. We acknowledge that the "what" of a commit message may overflow the summary line, into the body, and we prefer keeping the same syntactic form (often called the imperative; I'm on the fence about whether it's kind of a base infinitive, it would be good to figure that out 😆). @gnprice has posted some thoughts about this at https://github.com/zulip/zulip-mobile/pull/4079#issuecomment-622188734.

gnprice replied:
Note that as described in that comment, there's variation on this within the server project :slightly_smiling_face: -- some of us consistently use the same syntactic form as in the summary line, others consistently use the indicative (like "Renames"), and Tim varies between the two. But in the mobile repo we've generally been consistent in using the same form as in the summary line.

gnprice commented on src/webview/webViewEventHandlers.js:
```diff
@@ -278,6 +284,17 @@ export const handleMessageListEvent = (props: Props, _: GetText, event: MessageL
+      );
+      Alert.alert('', alertText);
+      break;
+    }
     default:
```
nit: blank line

gnprice commented on src/webview/webViewEventHandlers.js:
```diff
@@ -278,6 +284,17 @@ export const handleMessageListEvent = (props: Props, _: GetText, event: MessageL
+      const alertText = _.intl.formatMessage(
+        {
+          id: "This time is in your timezone. Original text was '{originalText}'.",
+          defaultMessage: "This time is in your timezone. Original text was '{originalText}'.",
```
This feels redundant, as it's just the same string as `id`. We don't use it in the definition of `_` -- we just pass `id` there. (I searched for `formatMessage` to look for examples.)

Can you take this `defaultMessage` line out, and does it all still work as expected?

Based on a quick look at the API docs:
https://formatjs.io/docs/react-intl/api/#formatmessage
https://formatjs.io/docs/react-intl/api/#message-formatting-fallbacks
they agree with our practice at `_`, that it should work.

agrawal-d replied:
It works.

gnprice replied:
Hmm, I see. I agree, then. Seems like rather a wart in that API.

gnprice commented on src/webview/static/base.css:
```diff
@@ -475,3 +475,9 @@ blockquote {
   height: 32px;
   fill: hsla(0, 0%, 100%, 0.75);
 }
+time{
```
nit: space before `{`

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
 };
 
-export default rewriteImageUrls;
+/* Rewrite the 'time' elements under the given root, inclusive, to make them more readable.
```
```suggestion
/**
 * Rewrite the 'time' elements under the given root, inclusive, to make them more readable.
```

This way it's marked as jsdoc.

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+ * 2. Adds 'original-text' attribrute, to show original text alert dialog on press.
+ */
+const rewriteTime = (element: Element | Document) => {
+  // Find the image elements to act on.
```
The copy-paste is showing :wink:

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+    const timeStamp = elem.getAttribute('datetime');
+    const text = elem.innerText;
+
+    // 'text' and 'timeStamp' are guaranteed to be present. The conditional is
+    // present only to satisfy Flow.
```
This kind of comment makes me ask: how do you know / why are they guaranteed? Or put another way: what do you know that Flow doesn't? That's the question I most want answered from a comment here.

There's a similar conditional above in this file which provides a nice example of answering that question:
```
    // Not `img.src`, because that getter absolutizes relative URLs itself.
    const actualSrc = img.getAttribute('src');

    // A missing `src` is invalid per the HTML spec.
    if (actualSrc == null) {
      return;
    }
```

OK, Flow doesn't know that fact. It *does* actually know that an `HTMLImageElement` must have a `src: string` -- but we avoided `img.src` for the reason mentioned in the other comment, and instead wrote `img.getAttribute('src')`, and Flow can't make the connection between the string passed to `getAttribute` and the type.

But it's not obvious to me that there's a good excuse like that here.

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+  );
+
+  timeElements.forEach(elem => {
+    const timeStamp = elem.getAttribute('datetime');
```
Is there a reason not to use the more direct `elem.datetime` here?

If you're copying from the example of `rewriteImageUrls`, note that there's a comment there explaining the reason we sadly can't use `img.src` at the corresponding spot there.

agrawal-d replied:
I was not aware that it's possible to refer to element-specific attributes directly.

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,46 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+ */
+const rewriteTime = (element: Element | Document) => {
+  // Find the image elements to act on.
+  const timeElements = [].concat(
```
When I hover over `elem` below, Flow reports the type:
> HTMLElement | HTMLTimeElement | any

I think that's the root of at least some of the need you're finding to add a conditional that should be impossible. In reality these elements should all be of type `HTMLTimeElement`.

Note how that's arranged in the other function:
```js
  const imageTags: $ReadOnlyArray<HTMLImageElement> = [].concat(
    element instanceof HTMLImageElement ? [element] : [],
    Array.from(element.getElementsByTagName('img')),
  );
```
with the type annotation. Try that here?

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +55,14 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+ * Modify relevant parts of the rendered DOM tree under the given root,
+ * inclusive, to ensure functionality.
+ *
+ * If no root element is specified, the entire document is considered.
```
This feature had been marked as "DEPRECATED". Is there a reason to drop that?

agrawal-d replied:
I did not know the correct meaning of deprecated. I thought it meant 'no longer used' when I removed it - because it was being used and so I thought that the "DEPRECATED" tag was wrong.

Just now I searched for the meaning, and it means "usable but regarded as obsolete and best avoided, typically because it has been superseded", so the comment is not incorrect. I'll revert it.

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -15,10 +15,9 @@ const inlineApiRoutes: RegExp[] = ['^/user_uploads/', '^/thumbnail$', '^/avatar/
- * entire document.
+ * Private; exported only for tests.
  */
-const rewriteImageUrls = (auth: Auth, element: Element | Document = document) => {
+export const rewriteImageUrls = (auth: Auth, element: Element | Document) => {
```
Instead of exporting this, let's just have the tests refer to the renamed `rewriteHTML`.

That way it's entirely an implementation detail of this module that its functionality is broken up into a couple of different functions. That seems like the most appropriate way to think of where its interface is.

agrawal-d commented:
Thanks for the review, @gnprice! I've made the changes you suggested.

gnprice commented on src/webview/js/__tests__/rewriteHtml.js:
```diff
@@ -1,10 +1,10 @@
 // @flow strict-local
 
-import rewriteImageUrls from '../rewriteImageUrls';
+import rewriteHTML from '../rewriteHtml';
```
Something's not quite right about this line :wink:

agrawal-d replied:
I can't figure out what's wrong! :P

chrisbobbe replied:
Ah, well spotted, Greg! I see it too—check the variable against the path, letter-by-letter. 🙂

agrawal-d replied:
Oh, got it now! Capitalization!

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +53,54 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+  );
+
+  timeElements.forEach(elem => {
+    // Present only because the Flow DOM libdef does not have a specific definition
+    // for `getElementsByTagName('time')`, so the typedef of `timeElements` is wider
+    // than it actually is.
+    if (!(elem instanceof HTMLTimeElement)) {
+      return;
+    }
```
Hmm, really? I see this in `dom.js` in my latest /tmp/flow/flowlib_* directory:
```js
  getElementsByTagName(name: 'time'): HTMLCollection<HTMLTimeElement>;
```

gnprice replied:
... Oho. It has that on `Document`... but not on `Element`.

This would be great to report on the Flow issue tracker :slightly_smiling_face: : https://github.com/facebook/flow/issues
Then the comment can just link to the issue.

agrawal-d replied:
Done. https://github.com/facebook/flow/issues/8450

gnprice commented on src/webview/js/rewriteHtml.js:
```diff
@@ -56,4 +53,54 @@ const rewriteImageUrls = (auth: Auth, element: Element | Document = document) =>
+    const timeStamp = elem.dateTime;
+    const text = elem.innerText;
+
+    // 'text' is guaranteed to be present, as the server always supplies it.
+    // The conditional is present only to satisfy Flow.
+    if (text == null) {
+      return;
+    }
```
Is this conditional actually needed? The code gets a lot shorter if it's taken out -- and simpler to think about, as there's one fewer path through the control flow.

At a quick experiment, I get no Flow errors if I just remove this.

gnprice commented on src/webview/webViewEventHandlers.js:
```diff
@@ -270,6 +276,18 @@ export const handleMessageListEvent = (props: Props, _: GetText, event: MessageL
       console.debug(props, event); // eslint-disable-line
       break;
 
+    case 'time': {
```
Let's not break up `debug` from `warn` and `error` -- as is, they form a nice sequence of closely related cases.

agrawal-d replied:
Oh yes! Sorry about this :(

agrawal-d commented:
Thanks for the review, @gnprice! I've pushed some changes.

I'm still confused by https://github.com/zulip/zulip-mobile/pull/4175#discussion_r463424159 :p

agrawal-d commented:
Fixed the capitalization error now :)

gnprice commented:
And merged! Thanks for the revisions, and for reporting that Flow issue :memo:
